packages/bsky/src/services/indexing/plugins/post.ts

```diff
@@ -99,19 +100,46 @@ const insertFn = async (
     }
     await db.insertInto('post_embed_record').values(embed).execute()
   }
-  // Thread index
+
+  // Thread indexing
+
+  // Concurrent, out-of-order updates are difficult, so we essentially
+  // take a lock on the thread for indexing, by locking the thread's root post.
   await db
+    .selectFrom('post')
+    .forUpdate()
+    .selectAll()
+    .where('uri', '=', post.replyRoot ?? post.uri)
+    .execute()
+
+  // Track the minimum we know about the post hierarchy: the post and its parent.
+  // Importantly, this works even if the parent hasn't been indexed yet.
+  const ancestors = await db
     .insertInto('post_hierarchy')
-    .values({
-      uri: post.uri,
-      ancestorUri: post.uri,
-      depth: 0,
-    })
-    .onConflict((oc) => oc.doNothing()) // Supports post updates
+    .values([
+      {
+        uri: post.uri,
+        ancestorUri: post.uri,
+        depth: 0,
+      },
+      ...(post.replyParent
```

Comment:
any chance we could do this with fewer operators outside of the db query? my brain starts to glaze over when i see this many symbols 😅